compiler/optimizer/DeadTreesElimination.cpp

```diff
@@ -930,34 +973,56 @@ int32_t TR::DeadTreesElimination::process(TR::TreeTop *startTree, TR::TreeTop *e
       {
       TR::Node *anchor = it->tree->getNode();
       TR::Node *load = anchor->getChild(0);
+      TR_ASSERT(anchor->getOpCodeValue() == TR::compressedRefs || load->getOpCode().isReadBar(),
+            "node n%dn appears when only compressedrefs or read barriers are expected when optimizing anchoring node\n", anchor->getGlobalIndex());
       if (load->getReferenceCount() > 1)
          continue;
 
-      // We can eliminate the indirect load immediately, but for the moment the
-      // subtree providing the base object has to be anchored.
-
-      TR::Node *heapBase = anchor->getChild(1);
-
-      TR::Node::recreate(anchor, TR::treetop);
-      anchor->setAndIncChild(0, load->getChild(0));
-      anchor->setChild(1, NULL);
-      anchor->setNumChildren(1);
-
-      if (!heapBase->getOpCode().isLoadConst())
+      /*
+       * transform from:
+       *
+       * anchor/treetop
+       *    xloadi/xrdbari
+       *       baseObjectNode
+       *    child1
+       *    ...
+       *
+       * to:
+       *
+       * treetop
+       *    baseObjectNode
+       * treetop
+       *    child1
+       *    ...
+       */
```

Comment:
Eliminating a load indirect is not as simple as saying it has no other uses - if the symref were volatile for example it creates a happens before edge that may be required to continue to exist. This doesn't really feel like a dead trees operation - there is a valid operation here and there needs to be some analysis/thought to remove it. What is your reasoning for doing this here?

Reply:
Thanks for pointing this out I confused the other treetop with real use.  This part of the code won't kick in for rdbar because it implies a rdbar is hanging under 2 treetops which is the case handled by trivial deadtree elimination already. Removing rdbar handling code here.